### PR TITLE
Home work 4

### PR DESCRIPTION
**Pull request by MikolaiR**
appearance of the first fragment
Screen

detail fragment appearance
Screenshot2

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -44,6 +45,9 @@ dependencies {
     androidTestImplementation 'androidx.test.ext:junit:1.1.2'
     androidTestImplementation 'androidx.test.espresso:espresso-core:3.3.0'
 
+    //DownloadImage
+    implementation 'com.github.bumptech.glide:glide:4.11.0'
+    implementation 'com.squareup.picasso:picasso:2.5.2'
```
Comment (emartynov) [inline app/build.gradle:50]: Why both?

File: app/src/main/java/com/example/androidacademyhomework/UI/MovieDetailsFragment.kt [hunk 2]
```diff
@@ -6,11 +6,32 @@ import android.view.LayoutInflater
...
+import androidx.recyclerview.widget.LinearLayoutManager
+import androidx.recyclerview.widget.RecyclerView
 import com.example.androidacademyhomework.R
+import com.example.androidacademyhomework.adapters.ActorsAdapter
+import com.example.androidacademyhomework.domain.MoviesDataSource
+import com.example.androidacademyhomework.models.Movie
 
 class MovieDetailsFragment : Fragment() {
 
+    private lateinit var textViewDetailFragmentAgeLimit: TextView
```
Comment (emartynov) [inline app/src/main/java/com/example/androidacademyhomework/UI/MovieDetailsFragment.kt:23]: You should read about fragment lifecycle and his view lifecycle. By this, you're making a memory leak!

File: app/src/main/java/com/example/androidacademyhomework/UI/MovieDetailsFragment.kt [hunk 3]
```diff
@@ -21,12 +42,56 @@ class MovieDetailsFragment : Fragment() {
...
+        textViewDetailFragmentMovieGenre = view.findViewById(R.id.textViewDetailFragmentMovieGenre)
+        ratingBarDetailsFragment = view.findViewById(R.id.ratingBarDetailsFragment)
+        textViewDetailFragmentReviews = view.findViewById(R.id.textViewDetailFragmentReviews)
+        textViewDetailFragmentStory = view.findViewById(R.id.textViewDetailFragmentStory)
+        recyclerForActors = view.findViewById(R.id.recyclerForActors)
+        initRecycler()
         view.findViewById<Button>(R.id.buttonViewBack).setOnClickListener {
             returnToMovieListFragment()
         }
+        movieId = args.movieId
```
Comment (emartynov) [inline app/src/main/java/com/example/androidacademyhomework/UI/MovieDetailsFragment.kt:56]: I don't see a necessity to store id - you can always get it from args

File: app/src/main/java/com/example/androidacademyhomework/UI/MovieDetailsFragment.kt [hunk 4]
```diff
@@ -21,12 +42,56 @@ class MovieDetailsFragment : Fragment() {
...
+
+    override fun onStart() {
+        super.onStart()
+        getMovieFromId(movieId)
+        updateDataMovieForTextView()
+        updateData()
+    }
+
+    private fun getMovieFromId(movieId:Int){
+       MoviesDataSource().getMovies().forEach { if (it.id == movieId) movie = it }
```
Comment (emartynov) [inline app/src/main/java/com/example/androidacademyhomework/UI/MovieDetailsFragment.kt:76]:
```suggestions
movie = MoviesDataSource().getMovies().find{ it.id == moviewId }
```

File: app/src/main/java/com/example/androidacademyhomework/UI/MovieListFragment.kt [hunk 5]
```diff
@@ -7,10 +7,19 @@ import android.view.View
...
+import androidx.recyclerview.widget.RecyclerView
 import com.example.androidacademyhomework.R
+import com.example.androidacademyhomework.adapters.MovieAdapter
+import com.example.androidacademyhomework.adapters.OnRecyclerItemClicked
+import com.example.androidacademyhomework.domain.MoviesDataSource
+import com.example.androidacademyhomework.models.Movie
 
 class MovieListFragment : Fragment() {
 
+    private var recyclerMovieList:RecyclerView? = null
```
Comment (emartynov) [inline app/src/main/java/com/example/androidacademyhomework/UI/MovieListFragment.kt:21]: Formatting!

File: app/src/main/java/com/example/androidacademyhomework/UI/MovieListFragment.kt [hunk 6]
```diff
@@ -20,13 +29,43 @@ class MovieListFragment : Fragment() {
...
+        recyclerMovieList?.apply {
+            layoutManager = GridLayoutManager(context, 2)
+            adapter = MovieAdapter(clickListener)
+            itemAnimator = DefaultItemAnimator()
+            hasFixedSize()
+        }
+    }
+
+    private val clickListener = object : OnRecyclerItemClicked {
+        override fun onClick(movie:Movie) {
```
Comment (emartynov) [inline app/src/main/java/com/example/androidacademyhomework/UI/MovieListFragment.kt:51]: Same about formatting

File: app/src/main/java/com/example/androidacademyhomework/UI/MovieListFragment.kt [hunk 7]
```diff
@@ -20,13 +29,43 @@ class MovieListFragment : Fragment() {
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
+        recyclerMovieList = view.findViewById(R.id.recyclerMovieList)
+        initRecycler()
+    }
+
+    override fun onStart() {
+        super.onStart()
+        updateData()
```
Comment (emartynov) [inline app/src/main/java/com/example/androidacademyhomework/UI/MovieListFragment.kt:38]: I'm not sure if you want to load data in adapter on every start. Please read about fragment lifecycle.

File: app/src/main/java/com/example/androidacademyhomework/UI/MovieListFragment.kt [hunk 8]
```diff
@@ -20,13 +29,43 @@ class MovieListFragment : Fragment() {
...
+    private fun updateData() {
+        (recyclerMovieList?.adapter as? MovieAdapter)?.apply {
+            bindActors(MoviesDataSource().getMovies())
         }
     }
 
-    private fun goToMovieDetailFragment(){
-        findNavController().navigate(R.id.action_movieListFragment_to_movieDetailsFragment)
+    override fun onDetach() {
+        recyclerMovieList = null
```
Comment (emartynov) [inline app/src/main/java/com/example/androidacademyhomework/UI/MovieListFragment.kt:63]: Super!!

File: app/src/main/java/com/example/androidacademyhomework/adapters/MovieAdapter.kt [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+            textViewMovieGenre.text = movie.genre
+            ratingBarListFragment.rating = movie.rating.toFloat()
+            textViewReviews.text = context.getString(R.string.reviews, movie.reviews)
+            textViewNameMove.text = movie.movieName
+            textViewFilmDuration.text = context.getString(R.string.duration_min,movie.movieDuration)
+        }
+    }
+}
+
+private val RecyclerView.ViewHolder.context
```
Comment (emartynov) [inline app/src/main/java/com/example/androidacademyhomework/adapters/MovieAdapter.kt:72]: You have this already in another adapter. Maybe to extract to common utility method?

File: app/src/main/java/com/example/androidacademyhomework/models/Actor.kt [hunk 10]
```diff
@@ -0,0 +1,7 @@
+package com.example.androidacademyhomework.models
+
+data class Actor (
+    val nameActor:String,
```
Comment (emartynov) [inline app/src/main/java/com/example/androidacademyhomework/models/Actor.kt:4]: Formatting and empty line

File: app/src/main/java/com/example/androidacademyhomework/models/Movie.kt [hunk 11]
```diff
@@ -0,0 +1,15 @@
+package com.example.androidacademyhomework.models
+
+data class Movie (
+    val id:Int,
```
Comment (emartynov) [inline app/src/main/java/com/example/androidacademyhomework/models/Movie.kt:4]: Also formatting

File: app/src/main/res/drawable-v24/error_image.xml [hunk 12]
```diff
@@ -0,0 +1,10 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (emartynov) [inline app/src/main/res/drawable-v24/error_image.xml:1]: Why in api 24?

File: app/src/main/res/drawable-v24/loading_animation.xml [hunk 13]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<animated-rotate xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (emartynov) [inline app/src/main/res/drawable-v24/loading_animation.xml:2]: Same :)

**Review by emartynov — COMMENTED** [review]
Some comments and looks good!

File: app/build.gradle (hunk 1 above)
Comment (MikolaiR) [reply app/build.gradle:50]: I was comparing the work of picasso and forgot to delete, sorry. Fixed

File: app/src/main/java/com/example/androidacademyhomework/UI/MovieDetailsFragment.kt (hunk 2 above)
Comment (MikolaiR) [reply app/src/main/java/com/example/androidacademyhomework/UI/MovieDetailsFragment.kt:23]: Fixed

File: app/src/main/java/com/example/androidacademyhomework/UI/MovieDetailsFragment.kt (hunk 3 above)
Comment (MikolaiR) [reply app/src/main/java/com/example/androidacademyhomework/UI/MovieDetailsFragment.kt:56]: You're right. Fixed

File: app/src/main/java/com/example/androidacademyhomework/UI/MovieDetailsFragment.kt (hunk 4 above)
Comment (MikolaiR) [reply app/src/main/java/com/example/androidacademyhomework/UI/MovieDetailsFragment.kt:76]: Yes, it really is better. Fixed

File: app/build.gradle (hunk 1 above)
Comment (emartynov) [reply app/build.gradle:50]: Cool, no sorry is needed

File: app/src/main/res/drawable-v24/error_image.xml (hunk 12 above)
Comment (MikolaiR) [reply app/src/main/res/drawable-v24/error_image.xml:1]: fixed

File: app/src/main/res/drawable-v24/loading_animation.xml (hunk 13 above)
Comment (MikolaiR) [reply app/src/main/res/drawable-v24/loading_animation.xml:2]: fixed

File: app/src/main/java/com/example/androidacademyhomework/adapters/MovieAdapter.kt (hunk 9 above)
Comment (MikolaiR) [reply app/src/main/java/com/example/androidacademyhomework/adapters/MovieAdapter.kt:72]: Fixed

**Comment by emartynov** [comment]
Can you push your changes?

**Pull request merged**